Python: extended notebook 06 to showcase Azure AI Search for semantic memory

rom212 opened the pull request:
### Motivation and Context

<!-- Thank you for your contribution to the semantic-kernel repo!
Please help reviewers and future users, providing the following information:
  1. Why is this change required?
  2. What problem does it solve?
  3. What scenario does it contribute to?
  4. If it fixes an open issue, please link to the issue here.
-->

Sample notebook 06-memory-and-embeddings demonstrated how semantic kernel lets users add memory to their kernels and also work around ever-growing contexts that eventually exceed the maximum token limit. It did so by using the in RAM VolatileMemoryStore and teased the possibility of offloading this to an external vector store. With this pull request, we want to provide guidance and repeatable steps so that users can actually perform semantic search by leveraging an external vector store, Azure AI Search (formerly known as Azure Cognitive Search

### Description

<!-- Describe your changes, the overall approach, the underlying design.
     These notes will help understanding how your code works. Thanks! -->

The changes consist in new cells added at the tail end of the existing notebook. We register a new memory store, external this time and show case how the rest of the code can be reused as-is. We add the capability of providing access credentials to Azure AI Search in environment variables as well.

### Contribution Checklist

<!-- Before submitting this PR, please make sure: -->

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

rom212 commented:
> @microsoft-github-policy-service agree [company="microsoft"]

moonbox3 reviewed (commented):
@rom212 Since you're already updating this notebook, can you please remove the unnecessary "a" in the following sentence in the intro text above the first cell?

Sentence to modify, please: "... would grow so large that you would run into ~~a~~ the model's token limit."

rom212 commented:
> @rom212 Since you're already updating this notebook, can you please remove the unnecessary "a" in the following sentence in the intro text above the first cell?
> 
> Sentence to modify, please: "... would grow so large that you would run into ~a~ the model's token limit."

Done. Sorry for missing it.

moonbox3 commented on python/notebooks/06-memory-and-embeddings.ipynb:
```diff
@@ -390,9 +515,126 @@
         "\n",
-        "Stay tuned for that!"
+        "azure_ai_search_api_key, azure_ai_search_url = sk.azure_aisearch_settings_from_dot_env()\n",
+        "kernel.register_memory_store(memory_store=AzureCognitiveSearchMemoryStore(1536,search_endpoint=azure_ai_search_url,admin_key=azure_ai_search_api_key))\n"
```
For the `vector_size` = 1536, could we use the named argument, please? It helps clarify for the reader/user if they're not familiar with what "1536" is.

moonbox3 replied:
My other small request with this line would be to try and create some line wrapping, please? Something like this:

kernel.register_memory_store(
    memory_store=AzureCognitiveSearchMemoryStore(
        vector_size=1536,
        search_endpoint=azure_ai_search_url,
        admin_key=azure_ai_search_api_key
    )
)

rom212 replied:
Hi Evan, I hope this is addressed in  commit 957f3d7d52d05075b1c524f32c5e037713dc73ae

moonbox3 commented on python/notebooks/06-memory-and-embeddings.ipynb:
```diff
@@ -390,9 +515,126 @@
+    },
+    {
+      "attachments": {
+        "image.png": {
```
I am having a difficult time reading the text in the images. Perhaps we could crop the image and show the most relevant part? This goes for the second image as well.

rom212 replied:
Hi Evan, I hope this is addressed to your satisfaction in commit 957f3d7d52d05075b1c524f32c5e037713dc73ae

moonbox3 commented on python/semantic_kernel/utils/settings.py:
```diff
@@ -230,3 +230,18 @@ def redis_settings_from_dot_env() -> str:
+    """
+    Reads the Azure AI Search environment variables for the .env file.
+
+    Returns:
```
I'd love to please include exactly what is returned (e.g., the Azure AI Search API Key and the Azure AI Search URL).

rom212 replied:
Makes sense as well, updated in commit 957f3d7d52d05075b1c524f32c5e037713dc73ae

moonbox3 commented:
Thank you for extending this notebook, @rom212. I suggest, after addressing the feedback, to commit the notebook without running all of the cells. All other example notebooks have been committed without running all.

moonbox3 commented:
> > @rom212 Since you're already updating this notebook, can you please remove the unnecessary "a" in the following sentence in the intro text above the first cell?
> > Sentence to modify, please: "... would grow so large that you would run into ~a~ the model's token limit."
> 
> Done. Sorry for missing it.

No worries at all! Thank you for doing it. Figured it'd be great to do the cleanup since you're already working with the file.

rom212 commented on python/semantic_kernel/utils/settings.py:
```diff
@@ -230,3 +230,18 @@ def redis_settings_from_dot_env() -> str:
+        dict: The Azure AI Search environment variables
+    """
+    config = dotenv_values(".env")
+    api_key = config.get("AZURE_AISEARCH_API_KEY", None)
```
Makes sense. added in commit 957f3d7d52d05075b1c524f32c5e037713dc73ae

rom212 commented:
> Thank you for extending this notebook, @rom212. I suggest, after addressing the feedback, to commit the notebook without running all of the cells. All other example notebooks have been committed without running all.

Ok Did a "restart" and "Clear All Outputs". Will keep this in mind.

moonbox3 commented:
Hi @rom212, can you please fix the following typos so that the proper checks can run successfully?

 "We can see that our function created an \"About Me\" index and that our five pieces of **<ins>infomation</ins>** have been indexed (note that it can take a few minutes for the UI to reflect the document count and storage size)."
 
 "We have laid the foundation which will allow us to store an **<ins>abitrary</ins>** amount of data in an external Vector Store above and beyond what could fit in memory at the expense of a little more latency."

Thanks!

rom212 commented:
> Hi @rom212, can you please fix the following typos so that the proper checks can run successfully?
> 
> "We can see that our function created an "About Me" index and that our five pieces of **infomation** have been indexed (note that it can take a few minutes for the UI to reflect the document count and storage size)."
> 
> "We have laid the foundation which will allow us to store an **abitrary** amount of data in an external Vector Store above and beyond what could fit in memory at the expense of a little more latency."
> 
> Thanks!

Done in dc8935141b9769b076f792abb8461e3742c3a8ef
Thanks!
